text/0000-nested-publish.md

```diff
@@ -0,0 +1,359 @@
+
+  This check is intended only to prevent accidents (such as vendoring a third-party package without considering the implications of redistributing it). It is always valid to omit `package.license` from the nested package, thus making no machine-readable claims about its licensing.
+
+It is an error for a nested package to have the same package name as the parent package or any other nested package with the same parent package. This is validated by all Cargo operations that would generate or read a lockfile. Rationale: This should ensure that whenever a nested package must be named, such as in an `.crate` archive, potentially in lock files, and potentially in Cargo user interface, the pair of (parent package name, nested package name) is sufficient to uniquely identify the package.
+
+However, it is allowed for a nested package to have the same name as a package that is _not_ a nested package with the same parent package (either because the latter has a different parent package or because it is not a nested package). This is appropriate because nested package names are an implementation detail of the package, and necessary to avoid different library packages from accidentally conflicting with each other by using the same nested package name.
+
+## **`cargo package` &amp; `cargo publish`**
+
+When a valid parent package is packaged, each of its transitive nested dependencies must be included in the `.crate` archive file. This has two sub-cases:
+
+* The nested package may be in a subdirectory of the parent package directory. In this case, it is copied to the same location in the archive, just like other packaged files.
+* Otherwise, it is copied to `.cargo/packages/<package name>/` within the archive.
+
+`Cargo.toml` files for nested packages are rewritten in the same way as is already done for all packages, except that `path` dependencies which are nested dependencies are kept, rather than stripped out or rejected, as they currently are. Their `path` values may need to be rewritten to point to the nested packages' new location in the archive.
+
+## **`crates.io`**
+
+`crates.io` will allow uploading of packages that contain `path` dependencies that were previously prohibited, as long as:
+
+* The dependency is a valid nested dependency as defined above. This includes that the the named package in fact exists in the `.crate` archive file, and has a valid `Cargo.toml` which declares `package.publish.nested = true`.
+* The `path`s in all contained manifests must not contain any upward traversal outside of the parent package (`../../`) or other hazardous or non-portable components as determined to be necessary.
+
+The package index, and the `crates.io` user interface, do not explicitly represent nested packages; the package is presented as if it were a single package:
+
+* Nested packages’ dependencies are flattened into the listed dependencies of the parent package.
```

Comment:
I'm a little concerned that this seems like it could be a fairly complex operation to flatten dependencies. For example, how does it handle name collisions? How does it handle fields that aren't normally unified, like the `public` field? I'm not quite sure I see how this could work.

If the intent is to have the resolver not know about nested packages, it seems like that would make it difficult for other parts of cargo to know about those nested packages since they are all driven by the resolver.

Reply:
Another field that can't be readily merged is `links`. Is it an error to have more than one `links` in the tree?

Reply:
If we don't hide this from the resolver, that opens up other areas of complexity

iiuc
- we would need a way for source ids to refer to these packages
- we would need a way to include these entries in the Index without conficting

Reply:
By merging dependencies we probably have to mean that the index will list all of the dependencies from the merged list. I.E. the list of dependencies will have more than one entry for `foo`. Rather than us merging all references to `foo` into one dependency. I Think, the resolver already has to deal with index entries with duplicate package names, for target specific dependencies and rename dependencies.

If we actually want to merge into one dependency entry, semver constraints are going to make life difficult. `>=1.0.0-alpha, <3.0.0` is not the merger of `>=1.0.0-alpha` with `<3.0.0`. (`1.0.0-alpha` matches the first two but not the third one).